src/pybind/mgr/volumes/fs/stats_util.py

```diff
@@ -57,3 +78,252 @@ def get_stats(src_path, dst_path, fs_handle, human=True):
+        Update the progress bar for ongoing + pending cloning operations.
+        '''
+        log.info('updating progress bars for clone jobs')
+        assert self.onpen_pev_id is not None
+
+        total_clones = len(clones)
+        percent = 0.0
+        avg_percent = 0.0
+
+        for clone in clones[total_clones:]:
+            with open_volume_lockless(self.volclient, clone.volname) as \
+                    fs_handle:
+                percent = get_percent_copied(clone.src_path, clone.dst_path,
+                                             fs_handle)
+                sum_percent += percent
+
+        avg_percent = round(sum_percent / total_clones, 3)
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        avg_progress_fraction = avg_percent / 100
+        msg = (f'{total_clones} ongoing+pending clones; Avg progress is '
+               f'{avg_percent}%')
+        self._update_progress_bar_event(ev_id=self.onpen_pev_id, ev_msg=msg,
+                                    ev_progress_fraction=avg_progress_fraction)
+
+        log.info('finished updating progress bar for ongong+pending clones')
+
+    def _update_ongoing_progress_bar(self, clones):
+        '''
+        Update the progress bar for ongoing cloning operations.
+        '''
+        log.info('updating progress bars for clone jobs')
+        assert self.on_pev_id is not None
+
+        percent = 0.0
+        sum_percent = 0.0
+        avg_percent = 0.0
+
+        if len(clones) < self.max_concurrent_clones:
+            total_clones = len(clones)
+        else:
+            total_clones = self.max_concurrent_clones
+
+        for clone in clones[:total_clones]:
+            with open_volume_lockless(self.volclient, clone.volname) as \
+                    fs_handle:
+                percent = get_percent_copied(clone.src_path, clone.dst_path,
+                                             fs_handle)
+                sum_percent += percent
+
+        avg_percent = round(sum_percent / total_clones, 3)
+        # progress module takes progress as a fraction between 0.0 to 1.0.
+        avg_progress_fraction = avg_percent / 100
+        msg = f'{total_clones} ongoing clones; Avg progress is {avg_percent}%'
+        self._update_progress_bar_event(ev_id=self.on_pev_id, ev_msg=msg,
+            ev_progress_fraction=avg_progress_fraction)
+
+        log.info('finished updating progress bar for ongoing clones')
+        return total_clones, sum_percent
+
+    def _update_progress_bars(self):
+        '''
+        Look for amount of progress made by all cloning operations and prints
+        progress bars, in "ceph -s" output, for average progress made
+        accordingly.
+
+        This method is supposed to be run only by instance of class RTimer
+        present in this class.
+        '''
+        clones = self._get_info_for_all_clones()
+
+        if clones:
+            total_clones, sum_percent = self._update_ongoing_progress_bar(clones)
```

Comment:
If size aggregation for both (ongoing and ongoing+pending) is done in one single function, returning these values isn't required.

Reply:
I know but have same function makes it little clutter-y and therefore I prefer current approach -- one task for method.

Reply:
It is relatively cluttery, I prefer the older version - 

```
    def _calculate_and_update_progress_bars(self, clones):
        '''
        Update the progress bar for ongoing cloning operations.
        '''
        log.info('calculation progress made by ongoing clones')
        assert self.on_pev_id is not None
        percent = 0.0
        sum_ongoing_percent = 0.0
        avg_ongoing_percent = 0.0
        total_ongoing_clones = min(len(clones), self.max_concurrent_clones)

        # onpen bar (that is progress bar for clone jobs in ongoing and pending
        # state) is printed when clones are in pending state. it is kept in
        # printing until all clone jobs finish.
        if len(clones) > self.max_concurrent_clones:
            log.info('calculation progress made by onpen clones')
            assert self.onpen_pev_id is not None
            sum_onpen_percent = 0.0
            avg_onpen_percent = 0.0
            total_onpen_clones = len(clones)

        for clone in clones:
            with open_volume_lockless(self.volclient, clone.volname) as \
                    fs_handle:
                percent = get_percent_copied(clone.src_path, clone.dst_path,
                                             fs_handle)
                if clone in clones[:total_clones]
                    sum_ongoing_percent += percent
                    sum_onpen_percent += percent
                elif len(clones) > self.max_concurrent_clones:
                    sum_onpen_percent += percent

        log.info('updating progress bars for ongoing clone jobs')
        avg_ongoing_percent = round(sum_oongoing_percent / total_clones, 3)
        # progress module takes progress as a fraction between 0.0 to 1.0.
        avg_ongoing_progress_fraction = avg_ongoing_percent / 100
        msg = f'{total_clones} ongoing clones; Avg progress is {avg_percent}%'
        self._update_progress_bar_event(ev_id=self.on_pev_id, ev_msg=msg,
            ev_progress_fraction=avg_ongoing_progress_fraction)
        log.info('finished updating progress bar for ongoing clones')

        if len(clones) > self.max_concurrent_clones:
            log.info('updating progress bars for onpen clone jobs')
            avg_onpen_percent = round(sum_onpen_percent / total_clones, 3)
            # progress module takes progress as a fraction between 0.0 to 1.0.
            avg_onpen_progress_fraction = avg_onpen_percent / 100
            msg = (f'{total_clones} ongoing+pending clones; Avg progress is '
                   f'{avg_percent}%')
            self._update_progress_bar_event(ev_id=self.onpen_pev_id, ev_msg=msg,
                                        ev_progress_fraction=avg_onpen_progress_fraction)
            log.info('finished updating progress bar for ongoing+pending clones')


```

Reply:
The earlier `if len(clones) > self.max_concurrent_clones:` isn't really required btw. And, really, you are not telling _why_ it is relatively cluttery. It makes total logical sense to iterate the list once and there is just a single conditional branch in the entire routine.

Reply:
> The earlier if len(clones) > self.max_concurrent_clones: isn't really required btw.

If you meant the first one, it is need due to the assert statement under it.

Reply:
>  And, really, you are not telling why it is relatively cluttery. 

I've got the same reason -- one task per method is a cleaner approach than 2 (or more tasks) per method. The code, the checks, the variables, the log entries, etc remains separate.

Reply:
> > And, really, you are not telling why it is relatively cluttery.
> 
> I've got the same reason -- one task per method is a cleaner approach than 2 (or more tasks) per method. The code, the checks, the variables, the log entries, etc remains separate.

I disagree. Normally, one task per method is fine, but, in this case its the same list that's iterated and there isn't much difference in the way aggregation is done, so why not just do that in once place.

> 
> > there is just a single conditional branch in the entire routine.
> 
> 3 actually. One for assert, one in loop and one for updating progress bars.

oh, you didn't the point. the additional branch off is minimal.

Sorry Rishabh - squashing functions to reduce code duplication with minimal branching seems like a no-brainer to me.

Reply:
In my opinion, 4-5 LoC is too minimal to mash. Anyways, that's only my opinion. Making the requested change.

Reply:
Done.